### ISSUE TITLE
Should the compiler always generate hash functions for records?

### ISSUE THREAD

**Issue by bmcdonald3** [chapel-lang/chapel#18577]
## Summary
My understanding is that currently the compiler generates a hash function for each record and this is done by hashing all of its fields and then combining those hashes together. This can cause the compiler-generated hash to not work in certain cases and cause the generated error to be difficult to decipher , such as when one of the fields is an `extern` type. It seems reasonable to either generate better error messages when the generated hash does not work or just not generate the hash function at all if a hash function does not exist on one of the record fields.

## Example code
The following code generates this error: `error: unresolved call '__mpz_struct.hash()'`, which seems strange since, to the user, it looks like we are trying to hash `bigint` and they would have no idea what `__mpz_struct` is or why it isn't working. 
```chpl
use BigInteger;
var D: domain(bigint);
```

## Proposed changes
Do not generate a compiler hash function if the record contains external types or is non-POD. For the case listed above, this would mean that the error would look something like: `error unresolved call 'bigint.hash()'`, instead of saying one of the fields hash function isn't defined. This would make it easier to decipher that defining a `bigint.hash()` would fix this issue.

### Related links
User issue: https://github.com/chapel-lang/chapel/issues/18562
PR to work around the compiler generated hash not working for this user issue: https://github.com/chapel-lang/chapel/pull/18574

**Comment by bmcdonald3** [comment]
Yes, thank you for pointing that out, updated!

**Comment by mppf** [comment]
> Do not generate a compiler hash function if the record contains external types or is non-POD. For the case listed above, this would mean that the error would look something like: error unresolved call 'bigint.hash()', instead of saying one of the fields hash function isn't defined. This would make it easier to decipher that defining a bigint.hash() would fix this issue.

I think we should always generate a hash function for records but when we do that, if one of the fields does not have a `.hash`, create a better error message to make that clearer. I'm open to not generating `bigint.hash` if hash is not available on one of the fields but I'm not so sure that this is better.

**Comment by mppf** [comment]
I was imagining the error would only appear if the hash function is used. `compilerError` could be used, or maybe we just find a way to make that error less confusing within the compiler.

**Comment by mppf** [comment]
I think the issue you are bringing up has more in common with `<` not being defined on all types than with POD-ness.

It's still my preference to automatically generate `<` as well as `hash` for user records but I can see your point about how having the hash use a pointer in an unexpected way could present problems.

**Comment by mppf** [comment]
Seems like maybe we should treat `hash` similarly to `<`. In PR #14752 I think we ended up with the situation that the auto-generated `<` will have a `where` clause returning `false` if `<` is not available on one of the fields. (But I haven't checked what this does today). Maybe we could/should use a similar approach.

**Comment by bradcray** [comment]
Hi Michael — I'm finally getting back to this.  I'm not really clear on how the "only create a hash function if all fields are hashable" suggestion works, in the sense that if a record contains a single field that's a class, it would pass this test, but that doesn't mean that the compiler-generated hash function would be semantically reasonable / useful, depending on the nature of the class?

**Comment by mppf** [comment]
That's a good point. We require that users who want to hash class instances opt in to doing so somehow. Then a record containing a class instance wouldn't be hashable (unless they had opted in). Another strategy is to distinguish between the `hash` function called by another `hash` function and the one called by the map/set/associative domain. That's kindof like what C++ did with comparison operators for `std::set` (we have `std::less` and `<` and they could be different).

**Comment by bradcray** [comment]
> We require...

Are you saying that we do already require this today, or is there a word missing (like "We could require").

> Another strategy is to distinguish between the hash function called by another hash function and the one called by the map/set/associative domain. 

I'm not understanding what this would buy us.  Is the idea that you'd have to explicitly enable the hash function for the map/set/associative domain, but would automatically get the other one?  And if so, what's the other one used/useful for?

**Comment by mppf** [comment]
Yes, I meant "we could require". Edited.

> I'm not understanding what this would buy us. Is the idea that you'd have to explicitly enable the hash function for the map/set/associative domain, but would automatically get the other one? And if so, what's the other one used/useful for?

We could separately choose whether a record containing a class is automatically hashable and whether a class itself is automatically hashable. I was imagining that we'd choose to make a class automatically hashable but not a record containing a class.

That was just an idea though. Probably a better idea to not generate hashes for classes at all though.

**Comment by bradcray** [comment]
Mulling this a bit over the holidays, I found myself realizing that I think most of the cases I've been focused on ("maybe these distinct class fields are equivalent" / "maybe these POD fields are actually indices into a table such that they're not merely POD") would also be cases where the user would effectively have to define their own `==` for their object in order to make it work correctly.  That made me wonder whether the rule here should be "If the record defines a `==` overload, the compiler should not introduce a `hash()` function.  This could effectively be the definition I was looking for in terms of the type being insufficiently POD above.

**Comment by bradcray** [comment]
Here's an example of a program that defines a record whose POD fields encode non-POD information that I think is problematic for creating a default hash function:

```chapel
use Set;

class C { var x: int; }

var myC = new C?(45);
var myC2 = new C?(33);

var A: [1..10] owned C?;

A[3] = myC;
A[5] = myC;
A[7] = myC2;

record R {
  var idx: int;
  operator ==(lhs: R, rhs: R) {
    return A[lhs.idx] == A[rhs.idx];
  }
}

var s: set(R);

var myR = new R(3);
var myR2 = new R(5);
var myR3 = new R(7);

s.add(myR);
s.add(myR2);
s.add(myR3);

writeln(s);
```

The situation is that the record looks POD in terms of its fields—they're  just integers.  But they're integers that index into a table of classes, and the `==` overload on the record is defining equality as "do their table entries point to the same thing or not?"  With the compiler-defined hash function, we store all three records (two of which are equal, the other of which is not) into the set and end up with a set of three entries.  Whereas if we took the presence of `==` as an indication that this record isn't trivially hashable, we'd just get a "can't create a set of R — no hash function" type of error which would be a cue to the user that they should probably define a hash function that takes the class identity into account.

I've got a branch that implements this idea if people agree.

**Comment by bradcray** [comment]
I agree that my use of POD here is a big / vague hammer and like the more precise terms you propose.  What I'm less sure of is whether there should be conditions other than "trivially comparable" that are taken into account when squashing the compiler-generated hash.  I.e., if a record is trivially comparable, but not trivially copyable, should we squash the hash?  Or are there other cases in which we should squash it?

**Comment by bradcray** [comment]
@mppf: In running testing on my branch, one of the (only) interesting failures is from your `test/exercises/duplicates/` hands-on exercises from a few years back.  Specifically, we create a map over SHA256Hash records, and SHA256Hash creates `==` and `!=` overloads.  I don't feel very conversant with this code, so was curious whether thinking about it in the context of this conversation made you think:
* "yeah, that's completely appropriate that a user would have to create a hash() function for it given that it defines those"
* "no, this changes my mind and I think we should consider some other rule."
* "it doesn't change my mind, but it's kindof a bummer to have to create one..."
* (something else)
Though I could easily be missing something, it appears to me as though it could define the <, <=, >, >= operators and not override the default == and != without a change in behavior (?).  OTOH, I also think that it could probably preserve the current behavior by defining:
```chapel
proc hash() {
  return hashVal.hash();
}
```
?

**Comment by bradcray** [comment]
> maybe you are focusing on == because that is what the hashtable needs to use

That's right.  In my current branch, if either `==` or `!=` are defined ("for symmetry" and because it seems like the common case), I squash the compiler-generated hash function.

**Comment by bradcray** [comment]
I think we resolved this issue in this discussion and implemented it in #19243, so am closing the issue.
